### PR TITLE
Modular subdag example

### PR DESCRIPTION
**Pull request by skrawcz**
Shows modular subdag example, and improves subdag config behavior.

## Changes
- example for modularity using subdag
- enables subdag config to handle `source` and `value`

## How I tested this
- locally
- unit tests

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: hamilton/function_modifiers/recursive.py [hunk 1]
```diff
@@ -423,7 +423,29 @@ def _derive_name(self, fn: Callable) -> str:
 
     def generate_nodes(self, fn: Callable, configuration: Dict[str, Any]) -> Collection[node.Node]:
         # Resolve all nodes from passed in functions
-        resolved_config = dict(configuration, **self.config)
+        # if self.config has source() or value() in it, we need to resolve it
+        sources_to_map = {}
```
Comment (elijahbenizzy) [inline hamilton/function_modifiers/recursive.py:427]: This should be it's own function `resolve_config` -- then we can test it directly

**Review by elijahbenizzy — COMMENTED** [review]
Two thoughts:
1. Break it out into a function to make it easier
2. The API is a bit odd -- `source` in the rest of Hamilton means that it's another node's output. Maybe we want to add a `config`/`from_config`? Or maybe a config mapping? Hmm...

File: hamilton/function_modifiers/recursive.py (hunk 1 above)
Comment (skrawcz) [reply hamilton/function_modifiers/recursive.py]: sure. done

File: examples/model_examples/modular_example/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+1. For training and predicting on the training set we use all 3 modules.
+2. For predicting on a separate dataset we use only the feature engineering module and the prediction module.
+3. We wire the two together so that the trained model then gets used in the prediction step for the separate dataset.
+
+By using `@subdag` we namespace the reuse of the modules and that's how we can
+reuse the same functions in different pipelines.
+
+See:
+![single_pipeline](my_dag.png)
```
Comment (elijahbenizzy) [inline examples/model_examples/modular_example/README.md:34]: Might be nice to highlight the two pipelines in this (E.G. by hand) to show what we're doing -- the naming doesn't make it obvious.

File: hamilton/function_modifiers/recursive.py [hunk 3]
```diff
@@ -131,6 +131,38 @@ def _validate_config_inputs(config: Dict[str, Any], inputs: Dict[str, Any]):
...
+    :param fields: the fields passed to the subdag decorator
+    :return: resolved configuration to use for this subdag.
+    """
+    sources_to_map = {}
+    values_to_include = {}
+    for key, value in fields.items():
+        if isinstance(value, dependencies.ConfigDependency):
+            sources_to_map[key] = value.source
+        elif isinstance(value, dependencies.LiteralDependency):
+            values_to_include[key] = value.value
```
Comment (elijahbenizzy) [inline hamilton/function_modifiers/recursive.py:149]: We should error out here if it's another type (E.G. `source`)

**Review by elijahbenizzy — COMMENTED** [review]
Nice work, two small nits

File: hamilton/function_modifiers/recursive.py (hunk 3 above)
Comment (skrawcz) [reply hamilton/function_modifiers/recursive.py:149]: done.

**Pull request merged**